bodhi/server/consumers/automatic_updates.py

```diff
@@ -0,0 +1,197 @@
+# Copyright © 2019 Red Hat, Inc. and others.
+#
+# This file is part of Bodhi.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+"""
+The Bodhi handler that creates updates automatically from tagged builds.
+
+This module is responsible for the process of creating updates when builds are
+tagged with certain tags.
+"""
+
+import logging
+import re
+import typing
+
+import fedora_messaging
+
+from bodhi.server import buildsys, initialize_db
+from bodhi.server.config import config
+from bodhi.server.exceptions import BodhiException
+from bodhi.server.models import Build, ContentType, Package, Release
+from bodhi.server.models import Update, UpdateType
+from bodhi.server.util import transactional_session_maker
+
+
+log = logging.getLogger('bodhi')
+
+
+class AutomaticUpdateHandler:
+    """
+    The Bodhi Automatic Update Handler.
+
+    A consumer that listens for messages about tagged builds and creates
+    updates from them.
+    """
+
+    _auto_update_cfg_key_re = re.compile(
+        r'(?P<releasename>\w+)\.autoupdates\.(?P<subkey>(?:\w+\.)*\w+)', re.ASCII)
+
+    _version_number_re = re.compile(r'[^\d]*(?P<vernum>\d+).*', re.ASCII)
+
+    def __init__(self, db_factory: transactional_session_maker = None):
+        """
+        Initialize the Automatic Update Handler.
+
+        Args:
+            db_factory (bodhi.server.util.TransactionalSessionMaker or None): If given, used as the
+                db_factory for this handler. If None (the default), a new TransactionalSessionMaker
+                is created and used.
+        """
+        if not db_factory:
+            initialize_db(config)
+            self.db_factory = transactional_session_maker()
+        else:
+            self.db_factory = db_factory
+
+    @classmethod
+    def _read_configuration(cls):
+        """Extract release <-> tags information from the configuration once."""
```

Comment:
I wonder if we should have this info available from the Release model object. Like we do for `mandatory_days_in_testing` (https://github.com/fedora-infra/bodhi/blob/develop/bodhi/server/models.py#L799).

Or we could directly get the config key for the release that is available in the msg body, any reason not to do that ?

Reply:
The purpose of this is to be able to find out the release from a tag into which a Koji build is tagged, it pulls that information for all configured releases/tags. I'm not sure how to feasibly do this in the context of a `Release` object, because we don't yet know which release it is when the message of a tagged build comes in. However, I could move this method and the related mappings to `Release` at the class level, then it's available if we need this info elsewhere. What do you think?

Reply:
I think we already have the Release - Koji tag mapping in the Release object. For example the candidate_tag (https://github.com/fedora-infra/bodhi/blob/develop/bodhi/server/models.py#L744) could be used. 
For example for the rawhide release the candidate tag would currently be f31 and we branch f31 the release can be updated to use the f32 tag.

We could find which release the build is associated with using this tag.